### PR TITLE
Improve TileMap physics for moving platforms and conveyor belts like movements

### PR DESCRIPTION
**Pull request by groud**
This allows both the TileMap node to act as a moving platform, or to define some tiles as a having a constant velocity (for conveyor belts). Also:
- Fixes a small bug when restoring the tile shape/snapping in the physics shape TileSet editor,
- ~~Fixes a bug with "texture_region_size" of atlas not being editable.~~ (fixed by https://github.com/godotengine/godot/pull/53095)

### CODE REVIEW

File: scene/2d/tile_map.cpp [hunk 1]
```diff
@@ -1051,18 +1062,26 @@ void TileMap::draw_tile(RID p_canvas_item, Vector2i p_position, const Ref<TileSe
...
 				for (int layer = 0; layer < (int)layers.size(); layer++) {
-					Transform2D global_transform = get_global_transform();
-
 					for (Map<Vector2i, TileMapQuadrant>::Element *E = layers[layer].quadrant_map.front(); E; E = E->next()) {
 						TileMapQuadrant &q = E->get();
 
 						Transform2D xform;
 						xform.set_origin(map_to_world(E->key() * get_effective_quadrant_size(layer)));
-						xform = global_transform * xform;
+						xform = new_transform * xform;
```
Comment (pouleyKetchoupp) [inline scene/2d/tile_map.cpp:1084]: It seems this line is causing the issues when sync to physics is disabled, it should use the global transform like before, since in this case `new_transform` is never updated.

Apart from that it seems to all work fine!

File: scene/2d/tile_map.cpp [hunk 2]
```diff
@@ -1163,8 +1205,7 @@ void TileMap::_physics_create_quadrant(TileMapQuadrant *p_quadrant) {
 	// Create the body and set its parameters.
 	for (int layer = 0; layer < tile_set->get_physics_layers_count(); layer++) {
 		RID body = PhysicsServer2D::get_singleton()->body_create();
-		PhysicsServer2D::get_singleton()->body_set_mode(body, PhysicsServer2D::BODY_MODE_STATIC);
-
+		PhysicsServer2D::get_singleton()->body_set_mode(body, PhysicsServer2D::BODY_MODE_KINEMATIC);
```
Comment (pouleyKetchoupp) [inline scene/2d/tile_map.cpp:1208]: Maybe the body type could be Kinematic only when sync to physics is enabled, otherwise by default changing the transform for the tilemap will collide with everything on the path.

File: scene/2d/tile_map.cpp [hunk 3]
```diff
@@ -34,6 +34,36 @@
...
+}
+
+Vector2i TileMapCollider::get_coords() const {
+	return coords;
+}
+
+void TileMapCollider::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_body"), &TileMapCollider::get_body);
+	ClassDB::bind_method(D_METHOD("get_tile_map"), &TileMapCollider::get_tile_map);
+	ClassDB::bind_method(D_METHOD("get_coords"), &TileMapCollider::get_coords);
```
Comment (pouleyKetchoupp) [inline scene/2d/tile_map.cpp:64]: The preferred way to access read-only values is usually properties with getter only instead of just exposing getter methods (like in `KinematicCollision2D`).

File: scene/2d/tile_map.cpp [hunk 4]
```diff
@@ -1070,24 +1112,67 @@ void TileMap::draw_tile(RID p_canvas_item, Vector2i p_position, const Ref<TileSe
 
 void TileMap::_physics_notification(int p_what) {
 	switch (p_what) {
+		case CanvasItem::NOTIFICATION_INTERNAL_PHYSICS_PROCESS: {
+			bool in_editor = false;
+#ifdef TOOLS_ENABLED
+			in_editor = Engine::get_singleton()->is_editor_hint();
+#endif
+			if (is_inside_tree() && collision_animatable && !in_editor) {
+				// Update tranform on the physics tick when in animatable name.
```
Comment (pouleyKetchoupp) [inline scene/2d/tile_map.cpp:1121]: Did you mean:
```suggestion
				// Update tranform on the physics tick when in animatable mode.
```

**Review by pouleyKetchoupp — APPROVED** [review]
Just left a couple nitpicks but it looks great and it's working fine (apart from the bug we already discussed and we're still investigating on the physics side).

**Edit:** Update about the character platform bug:
After investigation, it's due to changing the tilemap to use a separate body for each tile. CharacterBody2D applies the platform motion while ignoring the platform itself, but the way it works at the moment it can still collide with the next tile.
~It's most likely something that can be fixed in the physics system though, not a tilemap bug, so it's still all good :)~

**Edit 2:**
Actually this problem is a bit more complicated, because in order to handle platforms properly, we need all the tiles to either use the same body id, or get attached to the same object id and with this PR both of them is different for each tile.

@groud Would it be ok to change the design a bit in order to handle this case? Like attaching the Tilemap object itself to all tiles, and then have an API in the tilemap to retrieve the tile information (maybe from the body id).
I think that would also be better to be able to retrieve the TileMap itself from physics queries, which (unless I'm missing something) is not possible with the new design.
If needed we can discuss a bit more in details in the chat.

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Removing approval for now just so we can decide first if some changes are needed in the design for TileMapCollider (see https://github.com/godotengine/godot/pull/52724#pullrequestreview-759181300)

File: scene/2d/tile_map.cpp (hunk 3 above)
Comment (groud) [reply scene/2d/tile_map.cpp:64]: Is it ? I thought we wanted to get rid of read-only properties. I don't mind changing it but I'd rather be sure.

**Comment by groud** [comment]
> Actually this problem is a bit more complicated, because in order to handle platforms properly, we need all the tiles to either use the same body id, or get attached to the same object id and with this PR both of them is different for each tile.

I am not really sure I understand. I used to have the bug even when using the TileMap itself attached to all bodies. I guess what you mean is that going back to this situation would help you solve the problem in the physics server?

To explain my problem, I had to go for a one tile = one body approach because I needed to be able to define a constant velocity for each tile. Also from that, different tiles on the same TileMap may have different collisions layers/mask, so I needed also different bodies for them. And finally, I also split the TileMap into quadrants which had one body each (though we could go for a body with hundreds of shapes instead maybe, I am not sure).

So basically, both the different constant velocities, layers/mask and quadrants lead me to a having one body per tile.

I don't mind having an API in the TileMap to retrieve the collision information, but I am no sure about how I could do that. I would kind of need a way to index the collision that just happened, but I have not clue how to do it. Maybe a solution would be to store the tile coordinates in the KinematicCollider2D metadata? The ones you retrieve with `get_collider_metadata()`.

File: scene/2d/tile_map.cpp (hunk 3 above)
Comment (pouleyKetchoupp) [reply scene/2d/tile_map.cpp:64]: You're right! I thought we were aiming at using properties rather than methods in general, but I didn't know gdscript didn't handle read-only properties. So I will rather remove the ones in physics :)

**Comment by pouleyKetchoupp** [comment]
> I am not really sure I understand. I used to have the bug even when using the TileMap itself attached to all bodies. I guess what you mean is that going back to this situation would help you solve the problem in the physics server?

The issue used to happen between quadrants since they were also using separate bodies, and it's now happening more easily with one body per tile.
So it's true, the main issue is that I don't have a way to solve this bug anymore without being able to access the TileMap object itself.

> So basically, both the different constant velocities, layers/mask and quadrants lead me to a having one body per tile.

Makes sense! It's fine to use one body per tile, as long as the TileMap can be accessed somehow.

> I don't mind having an API in the TileMap to retrieve the collision information, but I am no sure about how I could do that. I would kind of need a way to index the collision that just happened, but I have not clue how to do it. Maybe a solution would be to store the tile coordinates in the KinematicCollider2D metadata? The ones you retrieve with `get_collider_metadata()`.

Good idea, shape metadata is also accessible in all physics queries so that would work well.

File: doc/classes/TileData.xml [hunk 5]
```diff
@@ -123,6 +137,22 @@
 				Sets the polygons count for TileSet physics layer with index [code]layer_id[/code].
 			</description>
 		</method>
+		<method name="set_constant_angular_velocity">
+			<return type="void" />
+			<argument index="0" name="layer_id" type="int" />
+			<argument index="1" name="velocity" type="float" />
+			<description>
+				Sets the constant angular velocity. This angular velocity is applied to objects colliding with this tile.
```
Comment (pouleyKetchoupp) [inline doc/classes/TileData.xml:145]: Just to be perfectly clear, this could explicitly say that it doesn't rotate the tile. Something like:
```suggestion
				Sets the constant angular velocity. This does not rotate the tile. This angular velocity is applied to objects colliding with this tile.
```

File: doc/classes/TileData.xml [hunk 6]
```diff
@@ -123,6 +137,22 @@
...
+			</description>
+		</method>
+		<method name="set_constant_linear_velocity">
+			<return type="void" />
+			<argument index="0" name="layer_id" type="int" />
+			<argument index="1" name="velocity" type="Vector2" />
+			<description>
+				Sets the constant linear velocity. This linear velocity is applied to objects colliding with this tile. This is useful to create conveyor belts.
```
Comment (pouleyKetchoupp) [inline doc/classes/TileData.xml:153]: And it could be something similar here:
```suggestion
				Sets the constant linear velocity. This does not move the tile. This linear velocity is applied to objects colliding with this tile. This is useful to create conveyor belts.
```

File: doc/classes/TileData.xml [hunk 7]
```diff
@@ -37,6 +37,20 @@
 				Returns how many polygons the tile has for TileSet physics layer with index [code]layer_id[/code].
 			</description>
 		</method>
+		<method name="get_constant_angular_velocity" qualifiers="const">
+			<return type="float" />
+			<argument index="0" name="layer_id" type="int" />
+			<description>
+				Returns the constant angular velocity.
```
Comment (pouleyKetchoupp) [inline doc/classes/TileData.xml:44]: And this could be more explicit too in order to avoid any confusion:
```suggestion
				Returns the constant angular velocity applied to objects colliding with this tile.
```

File: doc/classes/TileData.xml [hunk 8]
```diff
@@ -37,6 +37,20 @@
...
+			</description>
+		</method>
+		<method name="get_constant_linear_velocity" qualifiers="const">
+			<return type="Vector2" />
+			<argument index="0" name="layer_id" type="int" />
+			<description>
+				Returns the constant linear velocity.
```
Comment (pouleyKetchoupp) [inline doc/classes/TileData.xml:51]: Same here:
```suggestion
				Returns the constant linear velocity applied to objects colliding with this tile.
```

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Looks good! Just left some comments about details.

File: scene/2d/tile_map.h [hunk 9]
```diff
@@ -116,7 +116,7 @@ struct TileMapQuadrant {
 	List<RID> occluders;
 
 	// Physics.
-	List<RID> bodies;
+	Map<Vector2i, RID> bodies;
```
[the remark this answers is not in the document]
Comment (groud) [reply scene/2d/tile_map.h:119]: Yes, it's used when updating the bodies transform (see use of `body_element.key`)

[the remark this answers is not in the document]
Comment (pouleyKetchoupp) [reply scene/2d/tile_map.h:119]: I see, so it's used for the key/value pair but not for the lookup function.

In this case, is it possible to use only one map `Map<RID, Vector2i>` since you do use lookup from RID to coord but not the other way around, instead of having two maps?

File: doc/classes/TileData.xml [hunk 10]
```diff
@@ -37,6 +37,20 @@
 				Returns how many polygons the tile has for TileSet physics layer with index [code]layer_id[/code].
 			</description>
 		</method>
+		<method name="get_constant_angular_velocity" qualifiers="const">
+			<return type="float" />
+			<argument index="0" name="layer_id" type="int" />
+			<description>
+				Returns the constant angular velocity applied to objects colliding with this tile..
```
Comment (pouleyKetchoupp) [inline doc/classes/TileData.xml:44]: Double period:
```suggestion
				Returns the constant angular velocity applied to objects colliding with this tile.
```

File: doc/classes/TileData.xml [hunk 11]
```diff
@@ -37,6 +37,20 @@
...
+			</description>
+		</method>
+		<method name="get_constant_linear_velocity" qualifiers="const">
+			<return type="Vector2" />
+			<argument index="0" name="layer_id" type="int" />
+			<description>
+				Returns the constant linear velocity applied to objects colliding with this tile..
```
Comment (pouleyKetchoupp) [inline doc/classes/TileData.xml:51]: Same:
```suggestion
				Returns the constant linear velocity applied to objects colliding with this tile.
```

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]

File: scene/2d/tile_map.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (groud) [reply scene/2d/tile_map.h]: That makes sense, I updated the PR so that it retrieves the coords from the body instead.

**Review by pouleyKetchoupp — APPROVED** [review]
Looks perfect!

**Comment by akien-mga** [comment]
Thanks!
